### PR TITLE
[DEV-1970] Make deployment create/update a lock task

### PR DESCRIPTION
**Pull request by gee-senbong**
## Description

This PR 
* implements a `BaseLockTask` which can be used to run the task requires locking mechanism;
* makes deployment creation/update task a `BaseLockTask`.

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/worker/task/base.py [hunk 1]
```diff
@@ -99,3 +103,67 @@ async def execute(self) -> Any:
         """
         Execute the task
         """
+
+
+class BaseLockTask(BaseTask):
```
Comment (kchua78) [inline featurebyte/worker/task/base.py:108]: 👍

**Review by kchua78 — APPROVED** [review]

File: featurebyte/worker/task/deployment_create_update.py [hunk 2]
```diff
@@ -5,23 +5,43 @@
...
+
+    @property
+    def lock_blocking(self) -> bool:
+        # if the deployment is being created or updated by another task,
+        # fail the current task immediately without waiting
+        return False
+
+    def handle_lock_not_acquired(self) -> None:
+        payload = cast(DeploymentCreateUpdateTaskPayload, self.payload)
+        error_msg = f"Deployment {payload.output_document_id} is currently being created or updated"
```
Comment (YS-L) [inline featurebyte/worker/task/deployment_create_update.py:39]: nit: Maybe can customise the error message based on deployment payload type.

File: featurebyte/worker/task/base.py [hunk 3]
```diff
@@ -99,3 +103,67 @@ async def execute(self) -> Any:
...
+
+    async def execute(self) -> Any:
+        """
+        Execute the task
+
+        Returns
+        -------
+        Any
+        """
+        lock = self.app_container.redis.lock(self.lock_key)
```
Comment (YS-L) [inline featurebyte/worker/task/base.py:122]: Do we have to configure a lock timeout to handle unexpected scenarios where the lock is not released by abruptly terminated tasks?

Comment (gee-senbong) [reply featurebyte/worker/task/base.py:122]: Any suggestion for the lock expired time? If we add the lock at the task level, it is a bit difficult to set the expire value. That's why i did not include it in the current implementation.

File: featurebyte/worker/task/deployment_create_update.py (hunk 2 above)
Comment (gee-senbong) [reply featurebyte/worker/task/deployment_create_update.py:39]: We can't customize the error message based on the deployment payload type here unless we make another mongo call to retrieve the task payload (current type may not be the same as the one running on the other task).

File: featurebyte/worker/task/base.py (hunk 3 above)
Comment (YS-L) [reply featurebyte/worker/task/base.py:122]: Maybe we can set a large value like 24h as a default. The appropriate timeout could also be task dependent.

**Review by YS-L — APPROVED** [review]
LGTM

**Pull request merged**